src/wfs/src/main/java/org/geoserver/wfs/xml/FeatureTypeSchemaBuilder.java

```diff
@@ -365,20 +369,63 @@ public final XSDSchema buildSchemaInternal(
-            for (NamespaceInfo nameSpaceinfo : catalog.getNamespaces()) {
-                if (!schema.getQNamePrefixToNamespaceMap().containsKey(nameSpaceinfo.getPrefix())) {
-                    schema.getQNamePrefixToNamespaceMap()
-                            .put(nameSpaceinfo.getPrefix(), nameSpaceinfo.getURI());
+            if (localWorkspace != null && localWorkspace.isIsolated()) {
+                // Isolated namespace, add only required namespaces from mappings
+                final Map<String, String> schemaNamespacesMap =
+                        schema.getQNamePrefixToNamespaceMap();
+                final List<NamespaceInfo> catalogNamespaces = catalog.getNamespaces();
+                for (FeatureTypeInfo featureTypeInfo : featureTypeInfos) {
+                    Object mapObject =
+                            featureTypeInfo
+                                    .getFeatureType()
+                                    .getUserData()
+                                    .get(Types.DECLARED_NAMESPACES_MAP);
+                    if (!(mapObject instanceof Map)) continue;
+                    final Map<String, String> featureTypeNamespaces =
+                            (Map<String, String>) mapObject;
+                    featureTypeNamespaces
+                            .entrySet()
+                            .forEach(
+                                    entry -> {
+                                        final String uri = entry.getValue();
+                                        // check if URI is already taken
+                                        if (schemaNamespacesMap.containsValue(uri)) return;
+                                        // exists a prefix available in catalog for this URI?
+                                        final Optional<NamespaceInfo> nsFromCatalog =
+                                                catalogNamespaces
+                                                        .stream()
+                                                        .filter(
+                                                                nsi ->
+                                                                        Objects.equals(
+                                                                                nsi.getURI(), uri))
+                                                        .findFirst();
+                                        final String prefix =
+                                                nsFromCatalog
+                                                        .map(NamespaceInfo::getPrefix)
+                                                        .orElse(entry.getKey());
+                                        schemaNamespacesMap.put(prefix, uri);
+                                    });
+                }
+            } else {
+                // No isolated workspace, so add secondary namespaces from the full catalog
+                for (NamespaceInfo nameSpaceinfo : catalog.getNamespaces()) {
```

Comment:
This is wrong, the ``DECLARED_NAMESPACES_MAP`` shoudl already contain all necessary namespaces no? So why adding all the namespaces?

IMHO the behavior now should be, is a simple feature? business as usual. Is a complex feature? then we get the necessary namespaces from the complex type and we add them, makes sense?

Reply:
Indeed, makes sense.  Applied a change for now include all complex features use cases.